crates/pallets/starknet/src/transaction_validation.rs

```diff
@@ -50,3 +53,84 @@ impl<OuterOrigin: Into<Result<RawOrigin, OuterOrigin>> + From<RawOrigin>> Ensure
+                // the future queue of the transaction pool.
+                if sender_nonce < *transaction_nonce {
+                    log!(
+                        info,
+                        "Nonce is too high. Expected: {:?}, got: {:?}. This transaction will be placed in the \
+                         transaction pool and executed in the future when the nonce is reached.",
+                        sender_nonce,
+                        transaction_nonce
+                    );
+                }
+
+                (tx.sender_address(), sender_nonce, *transaction_nonce)
+            }
+            UserAndL1HandlerTransaction::L1Handler(tx, _fee) => {
+                let sender_address = ContractAddress::default();
+                let sender_nonce = Pallet::<T>::nonce(sender_address).into();
+                let nonce = tx.nonce;
+
+                Self::ensure_l1_message_not_executed(&Nonce(StarkFelt::from(nonce)))?;
+                (sender_address.into(), sender_nonce, nonce.into())
+            }
+        };
+
+        Ok((sender_address, sender_nonce, transaction_nonce))
+    }
+
+    pub fn validate_unsigned_tx(transaction: &UserAndL1HandlerTransaction) -> Result<(), InvalidTransaction> {
+        let chain_id = Self::chain_id();
+        let block_context = Self::get_block_context();
+        let mut state: BlockifierStateAdapter<T> = BlockifierStateAdapter::<T>::default();
+        let mut execution_resources = ExecutionResources::default();
+        let mut initial_gas = blockifier::abi::constants::INITIAL_GAS_COST;
+
+        match transaction {
+            UserAndL1HandlerTransaction::User(transaction) => {
+                match transaction {
+                    // There is no way to validate it before the account is actuallly deployed
+                    UserTransaction::DeployAccount(_) => Ok(None),
+                    UserTransaction::Declare(tx, contract_class) => tx
+                        .try_into_executable::<T::SystemHash>(chain_id, contract_class.clone(), false)
+                        .map_err(|_| InvalidTransaction::BadProof)?
+                        .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+                    UserTransaction::Invoke(tx) => tx.into_executable::<T::SystemHash>(chain_id, false).validate_tx(
+                        &mut state,
+                        &block_context,
+                        &mut execution_resources,
+                        &mut initial_gas,
+                        false,
+                    ),
+                }
+            }
+            UserAndL1HandlerTransaction::L1Handler(transaction, fee) => transaction
+                .into_executable::<T::SystemHash>(chain_id, *fee, false)
+                .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+        }
+        .map_or_else(|_error| Err(InvalidTransaction::BadProof), |_res| Ok(()))
```

Comment:
I'm nitpicking but


```rust
.map_err(|_| Err(InvalidTransaction::BadProof))?;

Ok(())
```

Is a more standard and therefore easily readable way of structuring this part of the code (IMO, maybe it's just me)